### PR TITLE
dart: 2.18.4 -> 2.19.3, refactor, and add tests

### PR DESCRIPTION
**Pull request by algorithmiker**
###### Description of changes

https://dart.dev/guides/whats-new#january-25-2023-219--30-alpha-releases
most notably smart enums, and `Isolate.run()`

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/dart/default.nix [hunk 1]
```diff
@@ -39,7 +40,7 @@
 assert version != null && version != "";
 assert sources != null && (builtins.isAttrs sources);
 
-stdenv.mkDerivation {
+stdenv.mkDerivation (final: {
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/dart/default.nix:43]:
```suggestion
stdenv.mkDerivation (finalAttrs: {
```

File: pkgs/development/compilers/dart/default.nix [hunk 2]
```diff
@@ -2,15 +2,16 @@
...
+, version ? "2.19.3"
 , sources ? let
     base = "https://storage.googleapis.com/dart-archive/channels";
     x86_64 = "x64";
     i686 = "ia32";
     aarch64 = "arm64";
     # Make sure that if the user overrides version parameter they're
     # also need to override sources, to avoid mistakes
-    version = "2.18.0";
+    version = "2.19.3";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/dart/default.nix:14]: This shouldn't really be an input

**Comment by GabrielDougherty** [comment]
Result of `nixpkgs-review pr 219875` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>dart</li>
  </ul>
</details>

**Comment by GabrielDougherty** [comment]
Result of `nixpkgs-review pr 219875` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>dart</li>
  </ul>
</details>

File: pkgs/development/compilers/dart/default.nix [hunk 3]
```diff
@@ -72,4 +91,4 @@ stdenv.mkDerivation {
     sourceProvenance = with sourceTypes; [ binaryNativeCode ];
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/dart/default.nix:91]: darwin breaks

```
Info: Compiling with sound null safety
Error: AOT compilation failed
ProcessException: No such file or directory
  Command: codesign -o linker-signed -f -s - /private/tmp/nix-build-dart-test-compile.drv-0/hello.exe
error: builder for '/nix/store/8zz4c6w3h9kjg81f93qmz40p10iz75py-dart-test-compile.drv' failed with exit code 64;
```

Comment (wegank) [reply pkgs/development/compilers/dart/default.nix:92]: Fixed.

File: pkgs/development/compilers/dart/default.nix [hunk 4]
```diff
@@ -56,9 +58,27 @@ stdenv.mkDerivation {
...
+    testCreate = runCommand "dart-test-create" { nativeBuildInputs = [ finalAttrs.finalPackage ]; } ''
+      PROJECTNAME="dart_test_project"
+      dart create --no-pub $PROJECTNAME
+
+      [[ -d $PROJECTNAME ]]
+      [[ -f $PROJECTNAME/bin/$PROJECTNAME.dart ]]
+      touch $out
+    '';
 
+    testCompile = runCommand "dart-test-compile" { nativeBuildInputs = [ finalAttrs.finalPackage ] ++ lib.optionals stdenv.isDarwin [ darwin.cctools darwin.sigtool ]; } ''
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/dart/default.nix:72]:
```suggestion
    testCompile = runCommand "dart-test-compile" {
      nativeBuildInputs = [ finalAttrs.finalPackage ]
        ++ lib.optionals stdenv.isDarwin [ darwin.cctools darwin.sigtool ];
    } ''
```

**Pull request merged**